### ISSUE TITLE
Embedded Support

### ISSUE THREAD

**Issue by Cylix** [thibault-martinez/iota.lib.cpp#252]
This issue is here to contain unfiltered search results about embedded C++ requirements.
Based on the resources we will post here, we then can determine what should be done for `v2.0.0` and work on it, as well as document it in the wiki.

**Comment by Cylix** [comment]
[MISRA C++ 2008](http://frey.notk.org/books/MISRA-Cpp-2008.pdf)

This document is considered as a reference for C++ secure programming.
It was originally intended for critical domains such as motor or aircraft industry, and extended to additional domains over the time.

This may give us very good insight about C++ secure programming for embedded devices if we aim to support such critical use cases in the future.

**Comment by Cylix** [comment]
**What C++ features should be avoided?**

From my current search results, one of the most important aspects of embedded development is the timing. On small devices, latency caused by language features is not tolerable.

When it comes to which features induce some cost and should be avoided, the most frequent answers are:
* exceptions, though this part is highly discussed
* RTTI (I don't think we do use that)
* dynamic memory allocation (this part is quite critical, considering we use std::string for example)
* virtual inheritance 

[Stackoverflow: What c++ features should be avoided for embedded development
 - 2010](https://stackoverflow.com/questions/3274582/what-c-features-should-be-avoided-for-embedded-development)

**Comment by Cylix** [comment]
**Compiler Support**

An important point to consider is that depending on the device, the device might provide a different C++ compiler with only a subset of the C++ features.
I guess it is unrealistic to support any kind of compiler, especially considering the diversity of devices.
However, it can be interesting to see what common features that kind of compilers are lacking.
Additionally, it can be interesting to test our build chain in such contexts.

**Comment by Cylix** [comment]
**Misc Resources**

[Information technology — Programming languages, their environments and system software interfaces — Technical Report on C++ Performance](http://standards.iso.org/ittf/PubliclyAvailableStandards/c043351_ISO_IEC_TR_18015_2006(E).zip)
[The Embedded C++](http://www.caravan.net/ec2plus/index.html)

**Comment by Cylix** [comment]
**EC++**

EC++ (Embedded C++) is a subset of C++ with the following features removed:
* multiple inheritances
* virtual base classes
* RTTI
* C++ cast style
* mutable
* namespaces
* exceptions
* templates

It was created at some early stage when there were few resources about C++ overhead.
It is mostly outdated and I guess we don't have to restrict ourselves so much.

From Bjarne himself: `To the best of my knowledge EC++ is dead (2004), and if it isn't it ought to be`
Only saw somewhere: `EC++ is a subset of C++, EC++ is only about enforcement of that subset; if you write code to the same subset using a C++ compiler, there is in fact zero additional overhead. The motivation for EC++ was largely to do with lack of a ratified C++ standard at the time of inception; since that no longer applies, EC++ is now largely irrelevant. The same subset can be enforces with static analysis tools without needing any special compiler or compiler mode.`

[Wikipedia](https://en.wikipedia.org/wiki/Embedded_C%2B%2B)

**Comment by Cylix** [comment]
**Some C++ Libraries with Embedded supports**

These libraries target embedded devices and may help us better understand what should be done.

* [embxx](https://github.com/arobenko/embxx/tree/master/embxx)

**Comment by Cylix** [comment]
**Dependencies**

We may need to think to diagnose our dependencies and see if they are appropriate. This includes
* cpr
* curl (subdependency of cpr)
* json
* keccak

**Comment by Cylix** [comment]
**Discussion about C++ Exceptions**

Speed should not be an issue anymore: there is no overhead for exception handling **if no exception occurs**.

The main issue reported is the impact on code side as it needs RTTI support or equivalent, which means more information embedded in the binary code. This may be an issue on some specific systems.

[Are Exceptions still undesirable in Realtime environment? (2011)](https://stackoverflow.com/questions/5257190/are-exceptions-still-undesirable-in-realtime-environment)

**Comment by Cylix** [comment]
**What need to be done**

This part is meant to keep track of element to consider for embedded support based on all the information provided in other comments.

It may (and most probably will) be updated, so this list is not definitive. It should instead be considered as a summary of the most important points.

* Reduce dynamic memory allocation footprint: limit as much as possible the number of dynamic memory allocations
* Investigate our use of exceptions. I don't think we need to drop exceptions, but I think we should restrict exceptions to make sure they do not happen frequently. A perfect example I can think of is the parallel_for in `bundlesFromAddresses` that does a loop of try/catch to dismiss invalid bundles: for our internal use, we may drop exception in that context. To address this part, we may need to list all places where we throw exceptions and all places where we catch some and take some decision.
* investigate cost of our json dependency and compare with `ArduinoJson` that sounds popular and fit for embedded.
* Strongly investigate cpr/curl and check for alternatives?
* Investigate our use of templates to see impact on code side

**Comment by Cylix** [comment]
**Analysis of libraries: cpr**

cpr is based on `libcurl`.
Some people report `libcurl` as too heavyweight, but some don't... check [this stackoverflow](https://stackoverflow.com/questions/23842394/c-c-http-client-library-for-embedded-projects).
The cpr wrapper does not seem too much heavyweight to be honest: it is mostly a C++ wrapper around the curl library.

My personal feeling though is that in our use case, we just need a very basic HTTP client that sends basic HTTP requests and receives basic responses.
We don't have to maintain sessions (cookies), we do not transfer files, we do not need other protocols, we do not even have to handle body encoding or compression technically. But all of those are example of features provided by curl, so my guess is that's why some people consider it heavyweight for embedded systems, especially for that kind of very basic use case.

If we'd like to go for something more lightweight, I think we would need to switch to something that does only what we want, but do it great (as of now, the main advantage of libcurl is that it is robust and probably fast).

Main thing we need to keep in mind is the portability aspect. So if we go for something lightweight, we still need to find something that is portable on different system, and that's probably the main limitation we will start to have.

The stackoverflow I linked previously suggests this [library](https://os.mbed.com/users/WiredHome/code/HTTPClient/shortlog/). Hard to tell if it is a good one or not considering they do not really use a standard platform to host the project..
It also lists this [one](https://os.mbed.com/users/WiredHome/code/HTTPClient/shortlog/), but it has not received any commit since 2013 (though, if it is working fine, it may not be necessary?).

Libcurl lists some alternatives on his [website](https://curl.haxx.se/libcurl/competitors.html)

To be honest, after going through this list and going around on google, I strongly feel libcurl is still our best shot for now. It is portable and robust. Most other libraries are either not maintained for long time, lack of users to prove they are reliable or performant, not portable or do still too many things.
Seems the C/C++ environment really sucks when it comes to HTTP libraries.

Also, we can also dig more libcurl. There are some configuration customization, such as [HTTP_ONLY](https://stackoverflow.com/questions/15251101/how-to-build-customize-libcurl-to-support-http-https-protocol-only) which is **not** set by cpr (cf [CMakeLists.txt](https://github.com/whoshuu/cpr/blob/master/CMakeLists.txt) and might be interesting to add). 
What we can do to go further would be to remove cpr and have a very raw wrapper around curl to make sure it is as lightweight as possible.

**Comment by Cylix** [comment]
**Analysis of libraries: json**

I think we should check [ArduinoJson](https://github.com/bblanchon/ArduinoJson).
It is highly starred and made for IoT with lots of (embedded) compilers support.
Main point is that it has **no dynamic memory allocation**, so it has very low memory footprint. This is done by having a static buffer for which we can change the size (so could be interesting to allow the library user to configure this setting too to fit as many devices/needs as possible).
Seems it mention it has no buffer copy and stream support, making it most probably fast (though we may need to compare with our current library).

As opposed, our current libraries does dynamic allocation, even though it tries to reduce the memory footprint to the minimum.

I found a very interesting comparison of dozens of C/C++ json libaries [here](https://github.com/miloyip/nativejson-benchmark#parsing-time). Interesting, ArduinoJson is not as good as our current library for speed and memory though? 
 
There is also this library used in an embedded OS: [frozen](https://github.com/cesanta/frozen).
Unfortunately, not tested in the above report.

However, in the report, [RapidJson](https://github.com/Tencent/rapidjson) always tops the speed and memory aspect. It is also developed and maintained by Tencent, the company behind WeChat.

We can also have a look to other libraries in the benchmark list, but seems that RapidJson is promising.

**Comment by Cylix** [comment]
**Analysis oflibraries: keccak**

I think we should not touch this library.
We are already using the original keccak libraries developed by their own creator. This is well maintained with wide support and optomization based on the platform.

What we would need to improve though is the configuration for each platform. Right now, we only distinguish between 32 and 64 bits while keccak can propose optimization for much more specific platforms.

**Comment by Cylix** [comment]
Spent some time benchmarking a bit rapidjson by doing 1000 GetNodeInfo.

Performance is same compared to our current json library.
However, as of now, rapidjson uses more memory than our current library.

Typically, our current library:
==256==     in use at exit: 168,960 bytes in 3,219 blocks
==256==   total heap usage: 229,978 allocs, 226,759 frees, 95,623,394 bytes allocated

rapidjson:
==2687==     in use at exit: 168,960 bytes in 3,219 blocks
==2687==   total heap usage: 183,001 allocs, 179,782 frees, 224,477,578 bytes allocated

rapidjson by reusing the input object to parse the output object:
==4730==     in use at exit: 168,960 bytes in 3,219 blocks
==4730==   total heap usage: 179,969 allocs, 176,750 frees, 158,876,322 bytes allocated

Note that reusing json object with our current library does not improve anything.
I'll check if I can get better memory usage.

**Comment by Cylix** [comment]
Just checked exceptions.

We mostly have exceptions for:
* Invalid model states
* Network failures
* Invalid parameters

Our exception usage is pretty much limited to corner cases and not general cases.

However, there are some places where our behavior is based on exceptions. Typical example would be the loop of try/verifyBundle/catch.

This, combined with the network failure based exception can be a potential limitation for embedded environment for which exceptions are disabled.
In such environment, the exception is replaced by an abort, making the library unsuitable/unusable in such environment.

If we are willing to support these environments, I guess the most simple way would be to follow the Google C++ guideline. I highly recommend these interesting readings:
* [Google C++ Guidelines](https://google.github.io/styleguide/cppguide.html#Exceptions)
* [Comment from Google engineer concerning this guideline and the STL](https://stackoverflow.com/a/15564254/1939604)
* [Google StatusOr<T> implementation in ProtoBuf](https://github.com/google/protobuf/blob/master/src/google/protobuf/stubs/statusor.h)
* [Quora and StatusOr<T>](http://qr.ae/TU1aGg)

To summarize, Google does not use exceptions, but still uses the STL (that does and thus will result in some abort).

I will make an experiment to see how it would look like to remove exceptions and have a google-style handling instead (the `StatusOr<T>` design).

However, to better integrate such change, I guess it would require to have a more standard way of returning values in Extended API. I will either think about #192 or make sure every function returns a response object to have some consistency.

**Comment by bluetiger9** [comment]
@thibault-martinez, did you got any progress on the ESP implementation? I see there is a [esp32](https://github.com/thibault-martinez/iota.lib.cpp/tree/esp32) branch in the repo, but has no commits.

**Comment by thibault-martinez** [comment]
@bluetiger9 Unfortunately, iota.lib.cpp is currently not under development anymore.
